Optimizer: support statically initialized `Atomic` and `Mutex` globals

eeckstein opened the pull request:
For example:
```
let global = Atomic<Int>(0)
```

This PR adds support for `@_rawLayout` structs and non-loadable types in InitializeStaticGlobals.

eeckstein committed:
SILGen: use builtin `prepareInitialization` instead of `zeroInitializer` when generating a value constructor for an empty non-copyable type

`zeroInitializer` is blocking optimizations. It is not needed because the constructor will initialize the memory anyway.
We only need to tell mandatory passes that this memory should be treated as initialized.

eeckstein committed:
SIL: add `rawLayoutSubstitutedLikeType` and `rawLayoutSubstitutedCountType` in `Type`

eeckstein committed:
SIL/AST: add `var InjectEnumAddrInst.element` and `var EnumElementDecl.hasAssociatedValues`

eeckstein committed:
DeadObjectElimination: handle the builtin "prepareInitialization"

which allows e.g. eliminating dead `Atomic` values

eeckstein committed:
InitializeStaticGlobals: allow statically initialized globals of non-loadable types

eeckstein committed:
InitializeStaticGlobals: support initializing globals with `@_rawLayout` types, like `Atomic`

Look through `@_rawLayout` projections, which "type casts" a raw-layout struct to it's content, which must match the like-type of the raw-layout, e.g.

```
@_rawLayout(like: T)
struct S {}

  %2 = builtin "addressOfRawLayout"<S>(%1 : $*S) : $Builtin.RawPointer
  %3 = pointer_to_address %2 to $*T
```

eeckstein committed:
InitializeStaticGlobals: support non-loadable enums

TODO: we don't support non-loadable enum cases with payload, yet, because IRGen support is missing.

Azoy reviewed (approved):
❤️ 🎉

Azoy commented:
It would also be nice to see if this patch allows statically initializing `Mutex` (on at least Darwin, Linux, and Windows)

eeckstein committed:
SIL: add the recursive property `isOrContainsRawLayout` to `SILTypeProperties`

eeckstein committed:
SIL: don't set the `let` flag on a SILGlobalVariable if it contains a struct with `@_rawLayout`

Even if the global declaration is a `let`.
Raw-layout storage may be mutated even for let-variables.
Treating such variables as non-let ensures that optimization don't assume that they are not mutated.

eeckstein committed:
IRGen: allow all kind of `Builtin.zeroInitializer` in statically initialized globals

Just use the same method to create the zero initializer constant as in `emitBuiltinCall`.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/InitializeStaticGlobals.swift:
```diff
@@ -376,17 +444,48 @@ private struct InitValueBuilder: AddressDefUseWalker {
+    case let injectEnum as InjectEnumAddrInst:
+      if injectEnum.element.hasAssociatedValues {
+        if !injectEnum.operand.value.type.isLoadable(in: injectEnum.parentFunction) {
+          // TODO: we don't support non-loadable enum cases with payload yet, because IRGen support is missing.
```
FixedTypeInfo (= used for non-loadable types) is missing the ability to pack a payload into an enum.
I added this in a comment.

eeckstein commented:
@swift-ci test
